server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MigratingMirrorToRepositoryService.java

```diff
@@ -0,0 +1,196 @@
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.centraldogma.common.Author;
+import com.linecorp.centraldogma.common.Change;
+import com.linecorp.centraldogma.common.Entry;
+import com.linecorp.centraldogma.common.Markup;
+import com.linecorp.centraldogma.common.Revision;
+import com.linecorp.centraldogma.internal.Jackson;
+import com.linecorp.centraldogma.server.command.Command;
+import com.linecorp.centraldogma.server.command.CommandExecutor;
+import com.linecorp.centraldogma.server.command.CommitResult;
+import com.linecorp.centraldogma.server.internal.replication.ZooKeeperCommandExecutor;
+import com.linecorp.centraldogma.server.internal.storage.repository.MirrorConfig;
+import com.linecorp.centraldogma.server.management.ServerStatus;
+import com.linecorp.centraldogma.server.storage.project.InternalProjectInitializer;
+import com.linecorp.centraldogma.server.storage.project.Project;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+import com.linecorp.centraldogma.server.storage.repository.MetaRepository;
+import com.linecorp.centraldogma.server.storage.repository.Repository;
+
+final class MigratingMirrorToRepositoryService {
+
+    private static final Logger logger = LoggerFactory.getLogger(MigratingMirrorToRepositoryService.class);
+
+    @VisibleForTesting
+    static final String MIRROR_TO_REPOSITORY_MIGRATION_JOB_LOG = "/mirror-to-repository-migration-job.json";
+
+    private final ProjectManager projectManager;
+    private final CommandExecutor commandExecutor;
+
+    MigratingMirrorToRepositoryService(ProjectManager projectManager, CommandExecutor commandExecutor) {
+        this.projectManager = projectManager;
+        this.commandExecutor = commandExecutor;
+    }
+
+    void migrate() throws Exception {
+        if (hasMigrationLog()) {
+            logger.debug("Mirrors have already been migrated to each repository. Skipping auto migration...");
+            return;
+        }
+
+        // Enter read-only mode.
+        commandExecutor.execute(Command.updateServerStatus(ServerStatus.REPLICATION_ONLY))
+                       .get(1, TimeUnit.MINUTES);
+        logger.info("Starting Mirrors migration ...");
+        if (commandExecutor instanceof ZooKeeperCommandExecutor) {
+            logger.debug("Waiting for 10 seconds to make sure that all cluster have been notified of the " +
+                         "read-only mode ...");
+            Thread.sleep(10000);
+        }
```

Comment:
I prefer migrating without entering read-only mode because
- It's not critical if mirroring is temporarily unavailable for a minute.
- Any issues can be promptly addressed if they arise.

If this is acceptable, I will remove these lines of code. Please let me know your opinion.

Reply:
Agreed.